File: claim-management-api/src/main/java/org/mavb/azure/ai/demos/service/ClaimService.java
package org.mavb.azure.ai.demos.service;

import org.mavb.azure.ai.demos.dto.request.*;
import org.mavb.azure.ai.demos.dto.response.*;
import org.springframework.http.codec.multipart.FilePart;
import reactor.core.publisher.Mono;

/**
 * Interfaz de servicio para la gestión de reclamos.
 * Define todos los métodos de negocio requeridos por el API.
 */
public interface ClaimService {

    /**
     * Crea un nuevo reclamo.
     *
     * @param createClaimDto DTO con los datos del reclamo a crear
     * @return ClaimDto con el reclamo creado
     */
    ClaimDto createClaim(CreateClaimDto createClaimDto);

    /**
     * Obtiene una lista paginada de reclamos con filtros opcionales.
     *
     * @param filterDto DTO con los filtros de búsqueda
     * @return ClaimListResponseDto con la lista de reclamos y paginación
     */
    ClaimListResponseDto getClaims(ClaimFilterDto filterDto);

    /**
     * Obtiene un reclamo específico por su ID.
     *
     * @param id ID del reclamo
     * @return ClaimDto con los datos del reclamo
     * @throws org.mavb.azure.ai.demos.exception.ClaimNotFoundException si el reclamo no existe
     */
    ClaimDto getClaimById(String id);

    /**
     * Resuelve un reclamo cambiando su estado a resuelto.
     *
     * @param id ID del reclamo a resolver
     * @param resolveClaimDto DTO con los comentarios de resolución
     * @return ClaimDto con el reclamo actualizado
     * @throws org.mavb.azure.ai.demos.exception.ClaimNotFoundException si el reclamo no existe
     */
    ClaimDto resolveClaim(String id, ResolveClaimDto resolveClaimDto);

    /**
     * Importa reclamos desde un archivo Excel usando WebFlux.
     *
     * @param filePart FilePart reactivo con el archivo Excel
     * @return Mono<ImportResponseDto> con el resultado de la importación
     * @throws org.mavb.azure.ai.demos.exception.InvalidFileException si el archivo no es válido
     */
    Mono<ImportResponseDto> importClaims(FilePart filePart);
}

File: showcase-website/src/app/modules/claim-management/services/claim.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { 
  Claim, 
  ClaimListResponse, 
  CreateClaimRequest, 
  ResolveClaimRequest, 
  ImportResponse, 
  ClaimFilters 
} from '../interfaces/claim.interface';
import { environment } from '../../../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class ClaimService {
  private http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiUrl}/claims`;

  /**
   * Obtiene lista paginada de reclamos con filtros opcionales
   */
  getClaims(filters: ClaimFilters = {}): Observable<ClaimListResponse> {
    let params = new HttpParams();
    
    if (filters.identityDocument) {
      params = params.append('identityDocument', filters.identityDocument);
    }
    if (filters.status) {
      params = params.append('status', filters.status);
    }
    if (filters.page) {
      params = params.append('page', filters.page.toString());
    }
    if (filters.limit) {
      params = params.append('limit', filters.limit.toString());
    }

    return this.http.get<ClaimListResponse>(this.baseUrl, { params });
  }

  /**
   * Obtiene un reclamo específico por ID
   */
  getClaimById(id: string): Observable<Claim> {
    return this.http.get<Claim>(`${this.baseUrl}/${id}`);
  }

  /**
   * Crea un nuevo reclamo
   */
  createClaim(claim: CreateClaimRequest): Observable<Claim> {
    return this.http.post<Claim>(this.baseUrl, claim);
  }

  /**
   * Marca un reclamo como resuelto
   */
  resolveClaim(id: string, resolution: ResolveClaimRequest): Observable<Claim> {
    return this.http.post<Claim>(`${this.baseUrl}/${id}/resolve`, resolution);
  }

  /**
   * Importa reclamos desde archivo Excel
   */
  importClaims(file: File): Observable<ImportResponse> {
    const formData = new FormData();
    formData.append('file', file);
    
    return this.http.post<ImportResponse>(`${this.baseUrl}/import`, formData);
  }
}